File: Programming Fundamentals/C#/Fundamentals/09.Regular Expressions/Exercise/P04_StarEnigma/P04_StarEnigma.cs
namespace P04_StarEnigma
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text.RegularExpressions;

    class P04_StarEnigma
    {
        static void Main(string[] args)
        {
            int n = int.Parse(Console.ReadLine());
            List<string> attackedPlanets = new List<string>();
            List<string> destroyedPlanets = new List<string>();

            for (int i = 0; i < n; i++)
            {
                string message = Console.ReadLine();
                string decrypted = string.Empty;
                MatchCollection matches = Regex.Matches(message, @"[STARstar]");

                for (int k = 0; k < message.Length; k++)
                {
                    decrypted += (char)(message[k] - matches.Count);
                }

                string pattern = @"@[^@\-!:>]*?(?<planet>[A-Z][a-z]+)[^@\-!:>]*?:[^@\-!:>]*?(?<population>\d+)[^@\-!:>]*?![^@\-!:>]*?(?<attackType>[AD])[^@\-!:>]*?![^@\-!:>]*?->[^@\-!:>]*?(?<soldiers>\d+)[^@\-!:>]*?";

                Match planet = Regex.Match(decrypted, pattern);
                if (planet.Groups["attackType"].Value == "A")
                {
                    attackedPlanets.Add(planet.Groups["planet"].Value);
                }
                else if (planet.Groups["attackType"].Value == "D")
                {
                    destroyedPlanets.Add(planet.Groups["planet"].Value);
                }                
            }

            Console.WriteLine($"Attacked planets: {attackedPlanets.Count}");
            foreach (string planet in attackedPlanets.OrderBy(x => x))
            {
                Console.WriteLine($"-> {planet}");
            }

            Console.WriteLine($"Destroyed planets: {destroyedPlanets.Count}");
            foreach (string planet in destroyedPlanets.OrderBy(x => x))
            {
                Console.WriteLine($"-> {planet}");
            }
        }
    }
}

File: Programming Fundamentals/JS/09.Regular Expressions/Exercise/p04_StarEnigma.js
function solve(input = []) {
  let attackedPlanets = new Set();
  let destroyedPlanets = new Set();

  let n = +input.shift();
  for (let i = 0; i < n; i++) {
    let message = input[i];
    let decrypted = '';

    let matches = message.match(/[STARstar]/g);
    for (let k = 0; k < message.length; k++) {
      decrypted += String.fromCharCode(message.charCodeAt(k) - (matches !== null ? matches.length : 0));
    }

    let planetMatch = decrypted.matchAll(/@[^@\-!:>]*?(?<planet>[A-Z][a-z]+)[^@\-!:>]*?:[^@\-!:>]*?(?<population>\d+)[^@\-!:>]*?![^@\-!:>]*?(?<attackType>[AD])[^@\-!:>]*?![^@\-!:>]*?->[^@\-!:>]*?(?<soldiers>\d+)[^@\-!:>]*?/g);
    for (let match of planetMatch) {
      let planet = match.groups['planet'];
      let attackType = match.groups['attackType'];

      if (attackType === 'A') {
        attackedPlanets.add(planet);
      } else if (attackType === 'D') {
        destroyedPlanets.add(planet);
      }
    }
  }

  attackedPlanets = new Set([...attackedPlanets].sort((a, b) => a.localeCompare(b)));
  destroyedPlanets = new Set([...destroyedPlanets].sort((a, b) => a.localeCompare(b)));

  console.log(`Attacked planets: ${attackedPlanets.size}`);
  attackedPlanets.forEach(x => console.log(`-> ${x}`));

  console.log(`Destroyed planets: ${destroyedPlanets.size}`);
  destroyedPlanets.forEach(x => console.log(`-> ${x}`));
}

// Don't copy the calling of the function in judge, you won't get any points, just the code above
solve(['2', 'STCDoghudd4=63333$D$0A53333', 'EHfsytsnhf?8555&I&2C9555SR']);
solve(['3', "tt(''DGsvywgerx>6444444444%H%1B9444", 'GQhrr|A977777(H(TTTT', 'EHfsytsnhf?8555&I&2C9555SR']);
